internal/handlers/bank_send_restriction_test.go

```diff
@@ -0,0 +1,117 @@
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "1000nonrestrictedmarker,1000restrictedmarker,999700000stake", addr1afterBalance, "addr1afterBalance")
+
+	// send restricted marker from account with transfer rights and no required attributes, expect success
+	sendRMarker := banktypes.NewMsgSend(addr1, addr2, sdk.NewCoins(sdk.NewInt64Coin(rMarkerDenom, 100)))
+	ConstructAndSendTx(tt, *app, ctx, acct1, priv, sendRMarker, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "1000nonrestrictedmarker,900restrictedmarker,999550000stake", addr1afterBalance, "addr1afterBalance")
+	addr2afterBalance := app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "100restrictedmarker,1000000000stake", addr2afterBalance, "addr2beforeBalance")
+
+	// send non restricted marker, expect success
+	sendRMarker = banktypes.NewMsgSend(addr1, addr2, sdk.NewCoins(sdk.NewInt64Coin(nrMarkerDenom, 100)))
+	ConstructAndSendTx(tt, *app, ctx, acct1, priv, sendRMarker, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "900nonrestrictedmarker,900restrictedmarker,999400000stake", addr1afterBalance, "addr1afterBalance")
+	addr2afterBalance = app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "100nonrestrictedmarker,100restrictedmarker,1000000000stake", addr2afterBalance, "addr2beforeBalance")
+
+	sendRMarker = banktypes.NewMsgSend(addr2, addr1, sdk.NewCoins(sdk.NewInt64Coin(nrMarkerDenom, 50)))
+	ConstructAndSendTx(tt, *app, ctx, acct2, priv2, sendRMarker, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "950nonrestrictedmarker,900restrictedmarker,999400000stake", addr1afterBalance, "addr1afterBalance")
+	addr2afterBalance = app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "50nonrestrictedmarker,100restrictedmarker,999850000stake", addr2afterBalance, "addr2beforeBalance")
+
+	stopIfFailed(tt)
+}
+
+func ConstructAndSendTx(tt *testing.T, app piosimapp.App, ctx sdk.Context, acct *authtypes.BaseAccount, priv cryptotypes.PrivKey, msg sdk.Msg, expectedCode uint32, expectedError string) {
+	encCfg := sdksim.MakeTestEncodingConfig()
+	fees := sdk.NewCoins(sdk.NewInt64Coin(sdk.DefaultBondDenom, int64(NewTestGasLimit())))
+	acct = app.AccountKeeper.GetAccount(ctx, acct.GetAddress()).(*authtypes.BaseAccount)
+	txBytes, err := SignTxAndGetBytes(NewTestGasLimit(), fees, encCfg, priv.PubKey(), priv, *acct, ctx.ChainID(), msg)
+	require.NoError(tt, err, "SignTxAndGetBytes")
+	res := app.DeliverTx(abci.RequestDeliverTx{Tx: txBytes})
+	require.Equal(tt, expectedCode, res.Code, "res=%+v", res)
+	if len(expectedError) > 0 {
+		require.True(tt, strings.Contains(res.Log, expectedError), fmt.Sprintf("error msg does not contain %s", expectedError))
+	}
```

Comment:
This is not what `require.True` is to be used for.

```suggestion
	if len(expectedError) > 0 {
		require.Contains(res.Log, expectedError, "DeliverTx result.Log")
	}	
```

That way, when it fails, the resulting test failure output includes the `res.Log` content which greatly helps identify why the test failed.

The only time to use `.True(...)` or `.False(...)` is when the thing you're testing outputs a boolean. Otherwise, there's always better ways to make your assertions.

Furthermore, you don't need the `fmt.Sprintf` for the test failure message. In fact, it's better without it because a) that delays the string generation until it's needed, and b) the line is shorter.

Basically, this: `require.Equal(t, a, b, "result from thing %v", someThing)` results in the same test output as `require.Equal(t, a, b, fmt.Sprintf("result from thing %v", someThing))`. The former only does a `Sprintf` if the test fails and that message is needed, though.